src/main/java/build/buf/protovalidate/CustomOverload.java

```diff
@@ -611,40 +656,583 @@ private static boolean validateURI(String val, boolean checkAbsolute) {
   }
 
   /**
-   * Validates if the input string is a valid IP prefix.
+   * Returns true if the string is a valid IP with prefix length, optionally limited to a specific
+   * version (v4 or v6), and optionally requiring the host portion to be all zeros.
    *
-   * @param prefix The input string to validate as an IP prefix.
-   * @param ver The IP version to validate against (0 for any version, 4 for IPv4, 6 for IPv6).
-   * @param strict If strict is true and host bits are set in the supplied address, then false is
-   *     returned.
-   * @return {@code true} if the input string is a valid IP prefix of the specified version, {@code
-   *     false} otherwise.
+   * <p>An address prefix divides an IP address into a network portion, and a host portion. The
+   * prefix length specifies how many bits the network portion has. For example, the IPv6 prefix
+   * "2001:db8:abcd:0012::0/64" designates the left-most 64 bits as the network prefix. The range of
+   * the network is 2**64 addresses, from 2001:db8:abcd:0012::0 to
+   * 2001:db8:abcd:0012:ffff:ffff:ffff:ffff.
+   *
+   * <p>An address prefix may include a specific host address, for example
+   * "2001:db8:abcd:0012::1f/64". With strict = true, this is not permitted. The host portion must
+   * be all zeros, as in "2001:db8:abcd:0012::0/64".
+   *
+   * <p>The same principle applies to IPv4 addresses. "192.168.1.0/24" designates the first 24 bits
+   * of the 32-bit IPv4 as the network prefix.
+   */
+  static boolean isIPPrefix(String str, long version, boolean strict) {
+    if (version == 6L) {
+      Ipv6 ip = new Ipv6(str);
+      return ip.addressPrefix() && (!strict || ip.isPrefixOnly());
+    } else if (version == 4L) {
+      Ipv4 ip = new Ipv4(str);
+      return ip.addressPrefix() && (!strict || ip.isPrefixOnly());
+    } else if (version == 0L) {
+      return isIPPrefix(str, 6, strict) || isIPPrefix(str, 4, strict);
+    }
+    return false;
+  }
+}
+
+final class Ipv4 {
```

Comment:
I think we should extract this to a separate file. If we leave it here we should make this a static class as it uses nothing from the outer class.

```suggestion
static final class Ipv4 {
```

Reply:
Which outer class do you mean? This seems OK to have in the same file since it's a package-private class? FWIW, this is also an example of consistency. Should the other impls move their classes to separate files too? I originally had them separated in Go, but we moved everything into `library.go` to keep it in one place.